BUG: Fix for scipy.special seterr, geterr, errstate

steppi opened the pull request:
#### What does this implement/fix?
<!--Please explain your changes.-->
This PR is an alternative to #20316. `scipy.special.seterr`, `scipy.special.geterr`, `scipy.special.errstate` currently do not work for ufuncs not implemented in the extension module `_ufuncs` (whose cython code is generated by `_generate_pyx.py`). This is because `sf_error.c` manages an error state in a static array, and currently each extension module has a separate copy of `sf_error.c` compiled in, when they actually all should be sharing this state. As identified by @izaid, a potential solution is to manage this state in a shared library. This PR differs from #20316 in that the shared library only manages the shared state and nothing else. I think it's best to go with the minimal set of changes for now in order to leave more options open for the future.

steppi committed:
Check in attempt to get this working

steppi committed:
Add -DSP_SPECFUN_ERROR

steppi committed:
Take out ignore of special seterr, errstate

steppi committed:
Add test for special sf_error alt ufunc machinery

steppi committed:
Add new sf_error test for alt machinery

steppi committed:
Add exception for symbol hiding of sf_error_state shared lib

steppi committed:
Add install_dir for sf_error_state shared lib

steppi committed:
Add install_rpath $ORIGIN where needed

steppi committed:
Add sf_error_state shared lib to test_public_api skiplist

steppi committed:
Minimize symbols exported + prefix with scipy

steppi commented:
@rgommers, the failing job is trying to build a wheel on windows. Are we running against a fundamental limitation like mentioned [here](https://github.com/mesonbuild/meson-python/blob/2daa991ce41b206e0a83065c8861646ac1821ae0/mesonpy/_rpath.py#L61-L64) or is there some secret sauce to get this to work because the shared library is being used only by extension modules that live alongside it in the same folder?

rgommers commented:
> I guess maybe something with `delvewheel` like here?

Yep, adding the path with `delvewheel` should work - it's also the recommended solution at https://mesonbuild.com/meson-python/reference/limitations.html#shared-libraries-on-windows. https://github.com/mesonbuild/meson-python/pull/551 may avoid the need for this in the future (I really need to test that).

steppi commented:
> > I guess maybe something with `delvewheel` like here?
> 
> Yep, adding the path with `delvewheel` should work - it's also the recommended solution at https://mesonbuild.com/meson-python/reference/limitations.html#shared-libraries-on-windows. [mesonbuild/meson-python#551](https://github.com/mesonbuild/meson-python/pull/551) may avoid the need for this in the future (I really need to test that).

Cool. Thanks. I’m setting up windows in a virtual machine so I can work this out locally.

steppi committed:
EMPTY: [skip circle]

steppi commented:
I found a solution which works locally on windows: extracting the contents of the wheel  to a temporary directory and taking `lib_sf_error_state.dll` out of the extracted wheel and putting it on the path delvewheel searches. Hopefully this works in the action too.

steppi commented:
Nice! That did it. It looks like I also need to update https://github.com/scipy/scipy/blob/main/tools/wheels/repair_windows.sh

izaid commented:
> Nice! That did it. It looks like I also need to update https://github.com/scipy/scipy/blob/main/tools/wheels/repair_windows.sh

Fantastic news! Glad you were able to sort this out (relatively) quickly.

steppi committed:
EMPTY: [wheel build]

lucascolley commented:
hmm that sounds like it might be related to gh-19627

steppi commented:
> hmm that sounds like it might be related to [gh-19627](https://github.com/scipy/scipy/pull/19627)

Yeah. New test that just needs tolerance bump

https://github.com/scipy/scipy/blob/53467c8f3cbd5af234625f9effb9a0fde88dfb8a/scipy/signal/tests/test_filter_design.py#L4252-L4257

steppi commented:
@rgommers, let me know if this solution looks OK.

rgommers reviewed (commented):
I had a closer look. This is more problematic than I expected:

1. The internal shared library not being able to be easily handled by `delvewheel` was unexpected (to me, in hindsight it's obvious). 
2. This makes it significantly harder for contributors on Windows to build SciPy; the OpenBLAS DLLs we need have special handling on Windows with `python dev.py --with-scipy-openblas`, which this PR doesn't have
3. This is going to break conda-forge builds on Windows too (and that of other redistributors). 
4. It's likely going to break with a future meson-python PR: https://github.com/mesonbuild/meson-python/pull/551

So far the bad news. Now the good news: I think with https://github.com/mesonbuild/meson-python/pull/551 and adding `scipy.special` to the DLL search path just like we do for vendored `libopenblas.dll`, I think we can avoid all these issues in one go.

That is one alternative. The other one is to store the shared state elsewhere, like on one of the existing existing modules (it can be retrieved from there by other extension modules when importing the private extension module via Python).

Let me also add for completeness (not blocking):

5. This isn't threadsafe (nor was the old global state); this needs handling when work starts on free-threaded CPython support (a.k.a. no-GIL).

rgommers commented:
We don't need to wait for that `meson-python` PR. This is the way to preload the DLL on Windows, preloading this new shared library should fix things: https://github.com/scipy/scipy/blob/bb92c8014e21052e7dde67a76b28214dd1dcb94a/tools/openblas_support.py#L239-L274

Try adding that to `special/__init__.py` and see if the `delvewheel` stuff can then be removed?

steppi commented on scipy/_lib/tests/test_public_api.py:
```diff
@@ -211,6 +211,7 @@ def is_unexpected(name):
 SKIP_LIST = [
     'scipy.conftest',
     'scipy.version',
+    'scipy.special.libsf_error_state'
```
It seems to be a false positive based on how the test works. I tried to put an underscore in front of it, but when creating a shared library named X, I couldn't figure out how to make the generated binary not be named libX.so (libX.dll on windows).

rgommers replied:
Ah okay, fine as is in that case.

steppi commented:
> Try adding that to `special/__init__.py` and see if the `delvewheel` stuff can then be removed?

Very nice. I'll try this out

steppi commented:
Thanks @rgommers! I was not expecting this to be so difficult either. I'll get to work on the fix you suggested after finishing up the AMOS complex issues. 

> 5\. This isn't threadsafe (nor was the old global state); this needs handling when work starts on free-threaded CPython support (a.k.a. no-GIL).

Hopefully we'll have all of the ufuncs in one extension module then, and we won't need to worry about this anymore.

steppi committed:
EMPTY: [skip ci]

steppi committed:
Manually load libsf_error_state.dll when on Windows

steppi commented on scipy/special/__init__.py:
```diff
@@ -861,3 +861,35 @@ def _get_include():
+    """  # noqa: E501
+    import os
+    if os.name == "nt":
+        try:
+            from ctypes import WinDLL
+            basedir = os.path.dirname(__file__)
+        except:
+            pass
```
The linter is displeased with the bare except. I'm not sure what the try/except block is for. I just copied it from:

https://github.com/scipy/scipy/blob/bb92c8014e21052e7dde67a76b28214dd1dcb94a/tools/openblas_support.py#L252-L258

Is it actually necessary? I can throw in a `noqa`. I suppose it has something to do with the comment

```python
 # convention for storing / loading the DLL from
                # numpy/.libs/, if present
```

the linter is displeased with the bare except. I could add `no

rgommers replied:
yes, good you put it back. this is quite fragile and I can't quite remember what can go wrong; prefer to change as little as possible.

steppi commented:
Huh. That worked locally on my Windows VM.

steppi commented:
OK, I get this same error locally now

```
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\hostedtoolcache\windows\Python\3.11.8\x64\Lib\site-packages\delvewheel\_dll_utils.py", line 403, in get_all_needed
    raise FileNotFoundError(f'Unable to find library: {dll_name}')
FileNotFoundError: Unable to find library: libsf_error_state.dll
repairing D:\a\scipy\scipy\dist\scipy-1.14.0.dev0-cp311-cp311-win_amd64.whl
finding DLL dependencies
Processing d:\a\scipy\scipy\dist\scipy-1.14.0.dev0-cp311-cp311-win_amd64.whl
```

The initial call to `delvewheel` is failing because it can't find `libsf_error_state.dll` on its search path.

steppi committed:
Ignore libsf_error_state.dll when calling delvewheel

steppi committed:
Bypass linter

steppi committed:
Copy over comment from openblas_support.py

steppi commented:
Ok, this seems to have worked. Thanks for the suggestion @rgommers!

rgommers committed:
MAINT: move DLL pre-loading in `special/__init__.py` up [wheel build]

It should be above the imports for the extension modules that need it.

Also some minor tweaks to code comments.

rgommers reviewed (approved):
All seems happy, let's give this a go! Thanks @steppi

rgommers commented:
And thanks @izaid as well for the help!
